core/src/test/java/brooklyn/location/basic/LocationConfigUtilsTest.java

```diff
@@ -21,77 +21,134 @@
         ConfigBag config = ConfigBag.newInstance();
         config.put(LocationConfigKeys.PRIVATE_KEY_DATA, "mydata");
         config.put(LocationConfigKeys.PRIVATE_KEY_FILE, SSH_PRIVATE_KEY_FILE);
         
-        String data = LocationConfigUtils.getPrivateKeyData(config);
+        OsCredential creds = LocationConfigUtils.getOsCredential(config).doKeyValidation(false);
+        Assert.assertTrue(creds.hasKey());
+        // warnings, as it is malformed
+        Assert.assertFalse(creds.getWarningMessages().isEmpty());
+
+        String data = creds.getPrivateKeyData();
         assertEquals(data, "mydata");
     }
-    
-    @Test(groups="Integration")
-    public void testPreferPubilcKeyDataOverFile() throws Exception {
+
+    @Test(expectedExceptions=IllegalStateException.class)
+    public void testInvalidKeyData() throws Exception {
+        ConfigBag config = ConfigBag.newInstance();
+        config.put(LocationConfigKeys.PRIVATE_KEY_DATA, "mydata");
+        
+        OsCredential creds = LocationConfigUtils.getOsCredential(config).doKeyValidation(false);
+        Assert.assertTrue(creds.hasKey());
+        Assert.assertFalse(creds.getWarningMessages().isEmpty());
+        
+        creds.checkNoErrors();
+    }
+
+    public void testPreferPublicKeyDataOverFileAndNoPrivateKeyRequired() throws Exception {
         ConfigBag config = ConfigBag.newInstance();
         config.put(LocationConfigKeys.PUBLIC_KEY_DATA, "mydata");
         config.put(LocationConfigKeys.PUBLIC_KEY_FILE, SSH_PUBLIC_KEY_FILE);
+        config.put(LocationConfigKeys.PRIVATE_KEY_FILE, "");
         
-        String data = LocationConfigUtils.getPublicKeyData(config);
+        OsCredential creds = LocationConfigUtils.getOsCredential(config);
+        String data = creds.getPublicKeyData();
         assertEquals(data, "mydata");
+        Assert.assertNull(creds.get());
+        Assert.assertFalse(creds.hasPassword());
+        Assert.assertFalse(creds.hasKey());
+        // and not even any warnings here
+        Assert.assertTrue(creds.getWarningMessages().isEmpty());
     }
     
-    @Test(groups="Integration")
-    public void testReadsPrivateKeyFileWithTildaPath() throws Exception {
+    @Test(groups="Integration")  // requires ~/.ssh/id_rsa
+    public void testReadsPrivateKeyFileWithTildePath() throws Exception {
         ConfigBag config = ConfigBag.newInstance();
-        config.put(LocationConfigKeys.PRIVATE_KEY_FILE, SSH_PRIVATE_KEY_FILE);
-        
-        String data = LocationConfigUtils.getPrivateKeyData(config);
+        config.put(LocationConfigKeys.PRIVATE_KEY_FILE, SSH_PRIVATE_KEY_FILE_WITH_TILDE);
+
+        // don't mind if it has a passphrase
+        String data = LocationConfigUtils.getOsCredential(config).doKeyValidation(false).get();
         assertTrue(data != null && data.length() > 0);
     }
     
-    @Test(groups="Integration")
+    @Test(groups="Integration")  // requires ~/.ssh/passphrase-id_rsa
```

Comment:
Instead of adding another external file dependency, can use re-use the `SshToolAbstractIntegrationTest.SSH_KEY_WITH_PASSPHRASE` (i.e. use `~/.ssh/id_rsa_with_passphrase` as the default)?